sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionPollerFn.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.hcatalog;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.io.hcatalog.HCatalogIO.Read;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.DoFn.UnboundedPerElement;
+import org.apache.beam.sdk.transforms.SerializableComparator;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableList;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Unbounded poller to listen for new partitions. */
+@UnboundedPerElement
+class PartitionPollerFn extends DoFn<Read, Read> {
```

Comment:
Can this one be replaced with a Watch transform, whose PollFn returns a complete list of current partitions? it scales quite well (if I remember correctly, e.g. a million elements was not a problem, as long as you encode them efficiently), and seems like it would save you a huge amount of code in this PR.

Reply:
Just curiosity question @jkff, does this mean then that we won't need an Unbounded SDF then? How do you tackle the case when new partitions appear?
(KafkaIO has an open ticket on detecting new partitions so I was wondering if the approach you propose could work there too).

Reply:
Yes, it does mean that we don't need an SDF here at all; and yes, this would likely work for KafkaIO too. Watch would poll for the current list of partitions, and emit new ones as they appear without re-emitting the already known ones, just like when we watch for new files.

It also means that the current PR would not be introducing any new SDFs, but instead merely reusing an existing one (Watch) - which is a very good thing and in line with SDF's position as enabling more composable/reusable IO APIs as opposed to people having to reimplement similar features in different IOs.

Reply:
Wow, that's super awesome, thanks for bringing this idea @jkff !

Reply:
@jhalaria For further clarity what @jkff is suggesting is to use `Watch` to follow a pattern like [`FileIO.Match.continously`](https://github.com/apache/beam/blob/2798d626507c7f5c8699096b3a188200ac64a299/sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java#L627) does to poll and discover the new partitions.